parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java

```diff
@@ -264,4 +254,31 @@ private List<Map<ColumnPath, ColumnChunkMetaData>> getColumnChunkMetadataForRowG
     }
     return listBuilder.build();
   }
+
+  private Expression findsResidual(
+      Expression expr,
+      Schema expectedSchema,
+      MessageType typeWithIds,
+      BlockMetaData rowGroup,
+      boolean caseSensitive) {
+    ParquetMetricsRowGroupFilter metricFilter =
+        new ParquetMetricsRowGroupFilter(expectedSchema, expr, caseSensitive);
+    Expression metricResidual = metricFilter.residualFor(typeWithIds, rowGroup);
+    if (metricResidual == Expressions.alwaysFalse() || metricResidual == Expressions.alwaysTrue()) {
+      return metricResidual;
+    }
+
+    ParquetDictionaryRowGroupFilter dictFilter =
+        new ParquetDictionaryRowGroupFilter(expectedSchema, metricResidual, caseSensitive);
+    Expression dictResidual =
+        dictFilter.residualFor(typeWithIds, rowGroup, reader.getDictionaryReader(rowGroup));
+    if (dictResidual == Expressions.alwaysFalse() || dictResidual == Expressions.alwaysTrue()) {
+      return dictResidual;
+    }
+
+    ParquetBloomRowGroupFilter bloomFilter =
+        new ParquetBloomRowGroupFilter(expectedSchema, dictResidual, caseSensitive);
+    return bloomFilter.residualFor(
+        typeWithIds, rowGroup, reader.getBloomFilterDataReader(rowGroup));
```

Comment:
I feel like it would be helpful to define a `ParquetCombinedFilter` or something like that which encapsulates the metrics row group filter, the dictionary filter, and the bloom filter. It would also implement the residualFor API and It should also make it easier to test that class, and the combined behavior of these filters.

Reply:
Added a ParquetCombinedFilter.